File: bootloader/router/patch.py
import sys
import os
import re

script_dir = os.path.dirname(os.path.abspath(__file__))

patch = [
    {
        'file': 'Core/Src/main.c',
        'search': re.compile(r'(USART1 Initialization Function[\s\S]*?USART_InitStruct\.BaudRate = )115200'),
        'replace': r'\1PORT0_BAUDRATE',
    },
    {
        'file': 'Core/Src/main.c',
        'search': re.compile(r'(USART2 Initialization Function[\s\S]*?USART_InitStruct\.BaudRate = )115200'),
        'replace': r'\1PORT1_BAUDRATE',
    },
    {
        'file': 'Core/Src/main.c',
        'search': re.compile(r'(PA0[\s\S]*?GPIO_InitStruct\.OutputType = )LL_GPIO_OUTPUT_PUSHPULL'),
        'replace': r'\1(PORT0_OC ? LL_GPIO_OUTPUT_OPENDRAIN : LL_GPIO_OUTPUT_PUSHPULL)',
    },
    {
        'file': 'Core/Src/main.c',
        'search': re.compile(r'(PA2[\s\S]*?GPIO_InitStruct\.OutputType = )LL_GPIO_OUTPUT_PUSHPULL'),
        'replace': r'\1(PORT1_OC ? LL_GPIO_OUTPUT_OPENDRAIN : LL_GPIO_OUTPUT_PUSHPULL)',
    },
]

# --- Apply patches ---
for p in patch:
    file_path = os.path.join(script_dir, p['file'])

    if not os.path.exists(file_path):
        print(f"ERROR: File not found: {p['file']}", file=sys.stderr)
        sys.exit(1)

    with open(file_path, 'r', encoding='utf-8') as f:
        content = f.read()

    if isinstance(p['replace'], list):
        matches = list(p['search'].finditer(content))

        if len(matches) != len(p['replace']):
            print(
                f"ERROR: Match count ({len(matches)}) does not equal replacement array length ({len(p['replace'])}) in file {p['file']}",
                file=sys.stderr,
            )
            sys.exit(1)

        # Build result without re-matching replaced content
        result = ''
        last_index = 0

        for match, replacement in zip(matches, p['replace']):
            result += content[last_index:match.start()]
            result += replacement
            last_index = match.end()

        result += content[last_index:]
        content = result

    else:
        if not p['search'].search(content):
            print(f"ERROR: No matches found in file {p['file']}", file=sys.stderr)
            sys.exit(1)

        content = p['search'].sub(p['replace'], content)

    with open(file_path, 'w', encoding='utf-8') as f:
        f.write(content)

    print(f"Patched: {p['file']}")

print('Done.')

File: bootloader/patch.mjs
import fs from 'fs';
import path from 'path';
import { pathToFileURL, fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// --- Load patch file ---
const patchFileArg = process.argv[2];

if (!patchFileArg) {
    console.error('ERROR: No patch file provided');
    process.exit(1);
}

const patchFilePath = path.resolve(__dirname, patchFileArg);

if (!fs.existsSync(patchFilePath)) {
    console.error(`ERROR: Patch file not found: ${patchFileArg}`);
    process.exit(1);
}

let patch;
try {
    const mod = await import(pathToFileURL(patchFilePath));
    patch = mod.patch;
} catch (e) {
    console.error(`ERROR: Failed to load patch file: ${e.message}`);
    process.exit(1);
}

if (!Array.isArray(patch)) {
    console.error('ERROR: Patch file must export an array named "patch"');
    process.exit(1);
}

// --- Apply patches ---
for (const p of patch) {
    const filePath = path.join(__dirname, p.file);

    if (!fs.existsSync(filePath)) {
        console.error(`ERROR: File not found: ${p.file}`);
        process.exit(1);
    }

    let content = fs.readFileSync(filePath, 'utf8');

    if (Array.isArray(p.replace)) {
        const searchFlags = p.search.flags.includes('g') ? p.search.flags : `${p.search.flags}g`;
        const searchRegex = new RegExp(p.search.source, searchFlags);
        const matches = [...content.matchAll(searchRegex)];

        if (matches.length !== p.replace.length) {
            console.error(
                `ERROR: Match count (${matches.length}) does not equal replacement array length (${p.replace.length}) in file ${p.file}`
            );
            process.exit(1);
        }

        // Build result without re-matching replaced content
        let result = '';
        let lastIndex = 0;

        matches.forEach((match, i) => {
            const start = match.index;
            const end = start + match[0].length;

            result += content.slice(lastIndex, start);
            result += p.replace[i];
            lastIndex = end;
        });

        result += content.slice(lastIndex);
        content = result;

    } else {
        const hasMatch = p.search.test(content);

        if (!hasMatch) {
            console.error(`ERROR: No matches found in file ${p.file}`);
            process.exit(1);
        }

        if (p.search.global) {
            p.search.lastIndex = 0;
        }

        content = content.replace(p.search, p.replace);
    }

    fs.writeFileSync(filePath, content, 'utf8');
    console.log(`Patched: ${p.file}`);
}

console.log('Done.');